man/anno.man
.TH ANNO %manext1% 2005-12-07 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
anno \- annotate nmh messages
.SH SYNOPSIS
.HP 5
.na
.B anno
.RB [ \-help ]
.RB [ \-version ]
.RI [ +folder ]
.RI [ msgs ]
.RB [ \-component
.IR field ]
.RB [ \-inplace " | " \-noinplace ]
.RB [ \-date " | " \-nodate ]
.RB [ \-draft ]
.RB [ \-append ]
.RB [ \-list ]
.RB [ \-delete ]
.RB [ \-number
.IR [ num|all ]]
.RB [ \-preserve " | " \-nopreserve ]
.RB [ \-text
.IR body ]
.ad
.SH DESCRIPTION
.B anno
manipulates header fields or
.I annotations
in messages.
Header fields consist of a field name and an optional field body
as defined by RFC 2822.
The
.B \-component
switch specifies the field name, and the
.B \-text
switch specifies the field body.
.PP
The messages are either the
.I msgs
in the named folder, or the draft if invoked with the
.B \-draft
switch.
.PP
Usually, annotation is performed by the commands
.BR dist ,
.BR forw ,
and
.BR repl ,
if they are given the
.B \-anno
switch.  This allows you to keep track of your distribution of,
forwarding of, and replies to a message.
.PP
By using
.BR anno ,
you can perform arbitrary annotations of your own.
Each message selected will be annotated with the lines
.PP
    field:\ date
    field:\ body
.PP
The
.B \-nodate
switch inhibits the date annotation, leaving only the
body annotation.
.PP
By default,
.B anno
prepends the annotations to the message.
Annotations are instead appended if the
.B \-append
switch is specified.
.PP
If a
.B \-component
.I field
is not specified when
.B anno
is invoked,
.B anno
will prompt the user for a field name.
.PP
The field specified must be a valid RFC 2822-style message field name,
consisting only of alphanumeric characters and dashes.
The body may consist of arbitrary text.
.PP
Normally
.B anno
does the annotation in place in order to preserve
any links to the message.  You may change this by using the
.B \-noinplace
switch.
.PP
The
.B \-list
switch produces a listing of the field bodies for header fields with
names matching the specified component, one per line.
The listing is numbered, starting at 1, if the
.B \-number
switch is also used.
A tab character separates the number and the field body.
The field body is treated as if it is a file name, and only the final
path name component is listed.
The complete field body is listed if the
.B \-text
switch is used; the argument to the
.B \-text
switch is required but ignored.
.PP
The
.B \-delete
switch removes header fields from messages.
The first header field whose name matches the component is deleted if
no other options are specified.
If the
.B \-text
switch is used in conjunction with the
.B \-delete
switch, the first header field whose name matches the component and
whose body matches the text is deleted.
The text is treated as if it was a file name; if it begins with a
slash, the entire field body must match the text, otherwise just the
last path name component of the field body must match.
If the
.B \-number
switch is used in conjunction with the
.B \-delete
switch, header field
.I num
whose name matches the component is deleted.
The number matches that which is produced by the
.B \-list
switch.
The special value
.B all
can be used for the number, and causes all components that match the
name to be deleted.
.PP
By default,
.B anno
changes the last-accessed and last-modified times on annotated messages
to the time of the annotation.
.B anno
preserves the original times if the
.B \-preserve
switch is used.
A matching
.B \-nopreserve
switch allows time preservation to be turned off, if enabled
in the profile.
.SH FILES
.TP 20
$HOME/.mh_profile
The user's profile.
.SH "PROFILE COMPONENTS"
.PD 0
.TP 20
Path:
To determine the user's nmh directory.
.TP
Current\-Folder:
To find the default current folder.
.PD
.SH "SEE ALSO"
.IR dist (1),
.IR forw (1),
.IR repl (1)
.SH DEFAULTS
.PD 0
.TP 20
+folder
The current folder.
.TP
msgs
The current message.
.TP
-inplace
.TP
-date
.PD
.SH CONTEXT
If a folder is given, it will become the current folder.  The first
message annotated will become the current message.


NEWS
#
# NEWS - Release notes for nmh 1.8
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been nearly five years since the last release of nmh, and there have
been a number of significant changes since that last release.  Long-time
MH and nmh users should read carefully the NOTABLE CHANGES section, as
there are some significant changes to nmh behavior.  Otherwise, please
see the README and INSTALL files for help on getting started with nmh.

This release is dedicated to Norman Z. Shapiro, co-designer of the MH
Message Handling System.  MH is the predecessor of nmh.  Norm was an
active supporter of nmh development until he passed away in October of
2021.  We are most grateful to Norm for his stewardship of MH and nmh.
    https://en.wikipedia.org/wiki/Norman_Shapiro

For news of future releases, subscribe to the low-volume
    https://lists.nongnu.org/mailman/listinfo/nmh-announce

---------------
NOTABLE CHANGES
---------------

- Support for Content-MD5 header fields, MIME content cache functionality,
  and the message/partial MIME type have been removed.
- Gmail OAuth2/XOAUTH support for desktop applications has been effectively
  dropped, so nmh no longer supports it.  nmh support for Gmail API access
  is experimental, please post to nmh-workers@nongnu.org if you'd like to
  help with test and development.
- repl(1) -convertargs now allows editing of the composition draft between
  translation and any encoding of text content.  Because encoding can wrap
  long lines, the use of a paragraph formatter has been removed from
  mhn.defaults.

------------
NEW FEATURES
------------

- The default editor has been changed from 'vi' to 'prompter', to align with
  historical practice and reduce packaging dependencies on external programs.
- A new -checkbase64 switch has been added to mhfixmsg(1).
- inc(1)/msgchk(1) now support STARTTLS for the POP protocol.
- All TLS-supported protocols now will send the SNI (server name indicator)
  TLS extension.
- A new mh-format function %(ordinal) has been implemented to output the
  appropriate ordinal suffix for numbers.  Example: "%(num 22)%(ordinal)"
  will output "22nd".
- show and mhl now decode more addresses in header fields.
- Added warning from all programs that read the profile if the profile
  contains a post entry, which is ignored, but does not contain a
  postproc entry.  In other words, if you get this warning and want
  to suppress it, your options include:
  1) Remove your post profile entry.
  2) Make your post profile entry a comment by prepending it with the #:
     comment indicator.
  3) Add a postproc entry that points to the post that you use.  That can
     be viewed with "mhparam postproc".
- scan(1) -file argument can be a Maildir directory.
- Updated mhn.defaults to prefer mpv(1) over xv(1) and replace mpeg_play(1),
  and to use it for all video types, not just video/mpeg.  And prefer all
  other searched-for pdf viewers over acroread(1).
- Added mhshow-suffix-video.mp4 to mhn.defaults, for use by mhshow(1) and
  send(1).
- Removed support from mhn.defaults for application/x-ivs and text/richtext.
- Changed interpretation of argument to mhfixmsg(1) -decodeheaderfieldbodies
  switch to specify character set of the decoded field bodies.
- repl(1) -convertargs now allows editing of the composition draft between
  translation and any encoding of text content.
- install-mh(1) now enables the mh-draft(5) draft folder facility.

-----------------
OBSOLETE FEATURES
-----------------

- The generation and verification of a Content-MD5 field has been removed
  without deprecation.  The related -check and -nocheck options now error.
- The MIME content cache functionality has been mostly non-functional since
  the development on nmh, and consequently all of the content caching code
  and related switches (-cache/-rcache/-wcache) have been removed.
- Support for generating and reassembling message/partial messages has been
  removed; it seems that this has been broken since 1.5 and there is very
  little support across MUAs.
- Marked Gmail OAuth2/XOAUTH support as being unsupported.
- Support for the MHPDEBUG environment variable was removed.  It was
  deprecated in nmh 1.7.  The pick(1) -debug switch replaced it.
- The 'libdir' mhparam(1) component was removed.  It was deprecated in
  nmh 1.7, when it was replaced by a new 'libexecdir' component.

---------
BUG FIXES
---------

- Fixed bcc to work with sendmail/pipe, and better documented that dcc
  doesn't work with it [Bug 55700].
- An -attendee switch has been added to mhical(1), for use when more than one
  (or zero) attendees match a user's mailbox.
- Fixed inc(1) and %(me) function escape to not obey Local-Mailbox profile
  component.
- Fixed source charset in mhfixmsg textcharset verbose output.
- Fixed mhfixmsg charset determination of content added with -reformat.
- Fixed file descriptor leak in mhfixmsg when run on multiple input files.
- Fixed mhfixmsg(1) -decodeheaderfilebodies to support mixed encoded/undecoded.
- Fixed memory corruption in post(1) and inc(1) when using XOAUTH2,
  with 4 or more entries in the oauth-authservice file.
- Added alias expansion to From: address for use by sendfrom.
- Removed extra space added before header field bodies by dist(1) to $mhdraft.
- Fixed display of iCalendar object with multiple VEVENTS using a VTIMEZONE.
- Fixed allowable encodings with MIME message types to get closer to RFC 2046.
- Detect other files regardless of backup prefix [Bug #49476].
- Copy if hard link by refile(1) or send(1) fails with EACESS [Bug #56575].

----------------------------------------------------------------------------

#
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time

uip/anno.c
/* anno.c -- annotate messages
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 *
 *	Four new options have been added: delete, list, number, and draft.
 *	Message header fields are used by the new MIME attachment code in
 *	the send command.  Adding features to generalize the anno command
 *	seemed to be a better approach than the creation of a new command
 *	whose features would overlap with those of the anno command.
 *
 *	The -draft option causes anno to operate on the current draft file
 *	instead of on a message sequence.
 *
 *	The -delete option deletes header elements that match the -component
 *	field name.  If -delete is used without the -text option, the first
 *	header field whose field name matches the component name is deleted.
 *	If the -delete is used with the -text option, and the -text argument
 *	begins with a /, the first header field whose field name matches the
 *	component name and whose field body matches the text is deleted.  If
 *	the -text argument does not begin with a /, then the text is assumed
 *	to be the last component of a path name, and the first header field
 *	whose field name matches the component name and a field body whose
 *	last path name component matches the text is deleted.  If the -delete
 *	option is used with the new -number option described below, the nth
 *	header field whose field name matches the component name is deleted.
 *	No header fields are deleted if none of the above conditions are met.
 *
 *	The -list option outputs the field bodies from each header field whose
 *	field name matches the component name, one per line.  If no -text
 *	option is specified, only the last path name component of each field
 *	body is output.  The entire field body is output if the -text option
 *	is used; the contents of the -text argument are ignored.  If the -list
 *	option is used in conjunction with the new -number option described
 *	below, each line is numbered starting with 1.  A tab separates the
 *	number from the field body.
 *
 *	The -number option works with both the -delete and -list options as
 *	described above.  The -number option takes an optional argument.  A
 *	value of 1 is assumed if this argument is absent.
 */

#include "h/mh.h"
#include "annosbr.h"
#include "sbr/m_name.h"
#include "sbr/getarguments.h"
#include "sbr/seq_setcur.h"
#include "sbr/seq_save.h"
#include "sbr/smatch.h"
#include "sbr/trimcpy.h"
#include "sbr/m_draft.h"
#include "sbr/m_convert.h"
#include "sbr/getfolder.h"
#include "sbr/folder_read.h"
#include "sbr/folder_free.h"
#include "sbr/context_save.h"
#include "sbr/context_replace.h"
#include "sbr/context_find.h"
#include "sbr/ambigsw.h"
#include "sbr/path.h"
#include "sbr/print_version.h"
#include "sbr/print_help.h"
#include "sbr/error.h"
#include "sbr/utils.h"
#include "sbr/done.h"
#include "sbr/m_maildir.h"
#include "sbr/globals.h"

#define ANNO_SWITCHES \
    X("component field", 0, COMPSW) \
    X("inplace", 0, INPLSW) \
    X("noinplace", 0, NINPLSW) \
    X("date", 0, DATESW) \
    X("nodate", 0, NDATESW) \
    X("text body", 0, TEXTSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \
    X("draft", 2, DRFTSW) \
    X("list", 1, LISTSW) \
    X("delete", 2, DELETESW) \
    X("number", 2, NUMBERSW) \
    X("append", 1, APPENDSW) \
    X("preserve", 1, PRESERVESW) \
    X("nopreserve", 3, NOPRESERVESW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(ANNO);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(ANNO, switches);
#undef X

/*
 * static prototypes
 */
static void make_comp (char **);


int
main (int argc, char **argv)
{
    bool inplace, datesw;
    int msgnum;
    char *cp, *maildir;
    char *comp = NULL, *text = NULL, *folder = NULL, buf[BUFSIZ];
    char **argp, **arguments;
    struct msgs_array msgs = { 0, 0, NULL };
    struct msgs *mp;
    bool	append;			/* append annotations instead of default prepend */
    int		delete = -2;		/* delete header element if set */
    char	*draft = NULL;	/* draft file name */
    int		isdf = 0;		/* return needed for m_draft() */
    bool	list;			/* list header elements if set */
    int		number = 0;		/* delete specific number of like elements if set */

    if (nmh_init(argv[0], true, true)) { return 1; }

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    append = list = false;
    inplace = datesw = true;
    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW:
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW:
		    die("-%s unknown", cp);

		case HELPSW:
		    snprintf (buf, sizeof(buf), "%s [+folder] [msgs] [switches]",
			invo_name);
		    print_help (buf, switches, 1);
		    done (0);
		case VERSIONSW:
		    print_version(invo_name);
		    done (0);

		case COMPSW:
		    if (comp)
			die("only one component at a time!");
		    if (!(comp = *argp++) || *comp == '-')
			die("missing argument to %s", argp[-2]);
		    continue;

		case DATESW:
		    datesw = true;
		    continue;
		case NDATESW:
		    datesw = false;
		    continue;

		case INPLSW:
		    inplace = true;
		    continue;
		case NINPLSW:
		    inplace = false;
		    continue;

		case TEXTSW:
		    if (text)
			die("only one body at a time!");
		    if (!(text = *argp++) || *text == '-')
			die("missing argument to %s", argp[-2]);
		    continue;

		case DELETESW:		/* delete annotations */
		    delete = 0;
		    continue;

		case DRFTSW:		/* draft message specified */
		    draft = "";
		    continue;

		case LISTSW:		/* produce a listing */
		    list = true;
		    continue;

		case NUMBERSW:		/* number listing or delete by number */
		    if (number != 0)
			die("only one number at a time!");

		    if (argp - arguments == argc - 1 || **argp == '-')
			number = 1;
		    
		    else {
			if (strcmp(*argp, "all") == 0)
			    number = -1;

			else if (!(number = atoi(*argp)))
			    die("missing argument to %s", argp[-1]);

			argp++;
		    }

		    delete = number;
		    continue;

		case APPENDSW:		/* append annotations instead of default prepend */
		    append = true;
		    continue;

		case PRESERVESW:	/* preserve access and modification times on annotated message */
		    annopreserve(1);
		    continue;

		case NOPRESERVESW:	/* don't preserve access and modification times on annotated message (default) */
		    annopreserve(0);
		    continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		die("only one folder at a time!");
            folder = pluspath (cp);
	} else
		app_msgarg(&msgs, cp);
    }

    /*
     *	We're dealing with the draft message instead of message numbers.
     *	Get the name of the draft and deal with it just as we do with
     *	message numbers below.
     */

    if (draft != NULL) {
	if (msgs.size != 0)
	    die("can only have message numbers or -draft.");

	draft = mh_xstrdup(m_draft(folder, NULL, 1, &isdf));

	make_comp(&comp);

	if (list)
	    annolist(draft, comp, text, number);
	else
	    annotate (draft, comp, text, inplace, datesw, delete, append);

	done(0);
	return 1;
    }

    if (!context_find ("path"))
	free (path ("./", TFOLDER));
    if (!msgs.size)
	app_msgarg(&msgs, "cur");
    if (!folder)
	folder = getfolder (1);
    maildir = m_maildir (folder);

    if (chdir (maildir) == -1)
	adios (maildir, "unable to change directory to");

    /* read folder and create message structure */
    if (!(mp = folder_read (folder, 1)))
	die("unable to read folder %s", folder);

    /* check for empty folder */
    if (mp->nummsg == 0)
	die("no messages in %s", folder);

    /* parse all the message ranges/sequences and set SELECTED */
    for (msgnum = 0; msgnum < msgs.size; msgnum++)
	if (!m_convert (mp, msgs.msgs[msgnum]))
	    done (1);

    make_comp (&comp);

    /* annotate all the SELECTED messages */
    for (msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++) {
	if (is_selected(mp, msgnum)) {
	    if (list)
		annolist(m_name(msgnum), comp, text, number);
	    else
		annotate (m_name (msgnum), comp, text, inplace, datesw, delete, append);
	}
    }

    context_replace (pfolder, folder);	/* update current folder  */
    seq_setcur (mp, mp->lowsel);	/* update current message */
    seq_save (mp);	/* synchronize message sequences */
    folder_free (mp);	/* free folder/message structure */
    context_save ();	/* save the context file         */
    done (0);
    return 1;
}

static void
make_comp (char **ap)
{
    char *cp, buffer[BUFSIZ];

    if (*ap == NULL) {
	fputs("Enter component name: ", stdout);
	fflush (stdout);

	if (fgets (buffer, sizeof buffer, stdin) == NULL)
	    done (1);
	*ap = trimcpy (buffer);
    }

    if ((cp = *ap + strlen (*ap) - 1) > *ap && *cp == ':')
	*cp = 0;
    if (!**ap)
	die("null component name");
    if (**ap == '-')
	die("invalid component name %s", *ap);
    if (strlen (*ap) >= NAMESZ)
	die("too large component name %s", *ap);

    for (cp = *ap; *cp; cp++)
	if (!isalnum ((unsigned char) *cp) && *cp != '-')
	    die("invalid component name %s", *ap);
}
